### DISCUSSION TITLE
New Stage Management View & Polls

### DISCUSSION THREAD

**Discussion by rashfael** [venueless/venueless#247] [Feature Preview & Feedback]
Welcome to our first feature preview, which is going to be a whole basket of related improvements!
_While the following text is written in the present tense, the new features presented here are not *yet* available for production use._

# New Stage Management View
![2021-06-15-23-17-22](https://user-images.githubusercontent.com/1093590/122125301-96483a00-ce30-11eb-8b4b-0fd952334e5b.png)
We've split out a lot of moderation features for stages into a brand new view, which only moderators and admins can access.
In the stage management view you can see the chat, questions and polls at once, provided you have a large enough screen.

On the normal, public stage, moderators now see the same things as any other attendee and attendee-centric actions like voting on questions and polls only available on the public stage and not in the management view, where you will always see the results, whether you voted first or not.

### Copy-pastable links to presentation mode
![2021-06-15-23-42-42](https://user-images.githubusercontent.com/1093590/122127419-6babb080-ce33-11eb-917c-88a9ff334b05.png)

From the stage management view you can now generate links to the presenation mode, which you can open in any other browser without logging in. Yes, that also means OBS. Presentation views (originally only for questions) are now also available for chat and polls and show the respective pinned question or poll.

Of course all presentation views have transparent backgrounds and polls and questions automatically resize to fill as much space as possible without getting blurry, which enables you to do stuff like this:

![2021-06-01-20-49-26](https://user-images.githubusercontent.com/1093590/122128077-6ef36c00-ce34-11eb-8418-1b476b7828e6.png)

*(please don't make the generated links public, since they do contain all your personal permissions, at least for now.)*

Another future feature, stage schedule control, will also finds its home in the stage management view.

# Polls!
![2021-06-15-23-39-07](https://user-images.githubusercontent.com/1093590/122128592-330cd680-ce35-11eb-9026-8b4818abff43.png)

A widely requested feature, **polls**, is almost ready. Let your audience vote on sets of answers, predefined by moderators.
From the new stage management view moderators can create polls and manage their lifecycle between draft, public vote, voting closed, pinned and archived.
Attendees can vote on public polls in the new `Polls` tab in the stage sidebar and view results on polls they already voted on:
![2021-06-15-23-22-28](https://user-images.githubusercontent.com/1093590/122129253-19b85a00-ce36-11eb-896e-e29904fd3a90.png)

We're starting out with single choice polls, but multiple choice polls will follow.

For those of you able to run your own venueless instance, you can check out the code here https://github.com/venueless/venueless/pull/236 and in the branch [room-manage](/venueless/venueless/tree/room-manage).

**Have any questions, comments or ideas for the previewed features?**
**What additional or differently presented information would you like to see in a moderator-specific view?**
**How would you like to be able to customize the presentation views for chat, questions or polls?**
**Comment below!**

**Comment by raphaelm** [comment]
Feedback on first test:

- It's hard to notice the button that switches to moderator view. It would at least need a tooltip, but I'd like something more visible like e.g. a switch as in our exhibitor editor more.
- The "Polls" tab and the "Polls" part of the moderator view show up even if the module isn't active in the room
- Moderator view: The "three-dot menu button" is always visible for questions, but only on hover for polls, this is a little confusing
- You can't "un-pin" a question or poll, is that intentional or a backend flaw or a frontend flaw?
- deleting a poll should probably ask for confirmation
- The UI allows you to vote for a closed poll (doesn't seem to actually work, though)
- When a new poll is opened, the user who is just watching the stream with the chat tab open won't notice at all
- Something is weird with the state management, might be a backend bug, I have an re-opened poll now that I can't vote for and haven't voted for before :thinking: 

During creating the first polls, I also encountered these:

```
vue.runtime.esm.js:1888 TypeError: Cannot read property 'filter' of null
    at o.sortedPolls (app.0edb6610.js:210)
    at ta.get (app.0edb6610.js:63)
    at ta.evaluate (app.0edb6610.js:63)
    at o.sortedPolls (app.0edb6610.js:63)
    at o.<anonymous> (app.0edb6610.js:63)
    at ta.get (app.0edb6610.js:63)
    at new ta (app.0edb6610.js:63)
    at o.e.$watch (app.0edb6610.js:63)
    at ga (app.0edb6610.js:63)
    at _a (app.0edb6610.js:63)
```
and
```
vue.runtime.esm.js:1888 TypeError: Cannot read property 'slice' of null
    at o.sortedPolls (Polls.vue:35)
    at ta.get (vue.runtime.esm.js:4479)
    at ta.evaluate (vue.runtime.esm.js:4584)
    at o.sortedPolls (vue.runtime.esm.js:4836)
    at o.<anonymous> (vue.runtime.esm.js:503)
    at ta.get (vue.runtime.esm.js:4479)
    at new ta (vue.runtime.esm.js:4468)
    at o.e.$watch (vue.runtime.esm.js:4946)
    at ga (vue.runtime.esm.js:4906)
    at _a (vue.runtime.esm.js:4888)
```

**Reply by rixx** [reply]
> You can't "un-pin" a question or poll, is that intentional or a backend flaw or a frontend flaw?

Backend flaw. I just added an `unpin` command for questions and polls (a bit cleaner than calling `pin` with `id: null` in my eyes, but I could be convinced otherwise if you two'd prefer that). Added in 2fd55dc and 78bd08d.

**Reply by rixx** [reply]
> Something is weird with the state management, might be a backend bug, I have an re-opened poll now that I can't vote for and haven't voted for before

Interestingly, the backend didn't even validate that a poll was still open before accepting a vote, changing that now.

**Reply by rashfael** [reply]
> * Moderator view: The "three-dot menu button" is always visible for questions, but only on hover for polls, this is a little confusing
Yeah, the existing question view code wasn't in line with the other stuff where we show icons only on hover, and I haven't made that consistent yet.

**Reply by rashfael** [reply]
> * When a new poll is opened, the user who is just watching the stream with the chat tab open won't notice at all
This was caused by the race condition errors you posted wrecking the watcher for the red dots in the tabs. Working now.

**Comment by saschafoerster** [comment]
I would like to test the poll-view. I am using the latest version on docker but not the development-version. I saw, that I can activate polls in the user-rights-management-view, but then I don't find it in the room settings.

**Reply by raphaelm** [reply]
Its a feature flag you need to activate for the world in the /control/ interface. If you want it enabled on hosted venueless, just shoot over an emailw ith the world name

**Reply by saschafoerster** [reply]
@raphaelm Oh, plenty of new flags, I found them now! Thank you! :)
